cleanup/cleanup.go

```diff
@@ -0,0 +1,150 @@
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+	"github.com/grpc/test-infra/config"
+	pb "github.com/grpc/test-infra/proto/grpc/testing"
+	"github.com/grpc/test-infra/status"
+)
+
+// CleanupAgent
+
+// +kubebuilder:rbac:groups=e2etest.grpc.io,resources=loadtests,verbs=get;list;watch
+// +kubebuilder:rbac:groups=e2etest.grpc.io,resources=loadtests/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=get
+
+// Agent cleanup unwanted processes.
+type Agent struct {
+	client.Client
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	Timeout time.Duration
+}
+
+// Reconcile attempts to check status of workers of the triggering LoadTest, if
+// a terminated LoadTest has workers still running, reconcile will send quit RPC
+// to stop the workers.
+func (a *Agent) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var ctx context.Context
+	var cancel context.CancelFunc
+	var err error
+
+	// The timeout for the cleanup process could be set by maintainer, but if not
+	// set the whole cleanup process is bonded by 2 mins.
+	if a.Timeout == 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(120)*time.Second)
+	} else {
+		ctx, cancel = context.WithTimeout(context.Background(), a.Timeout)
+	}
+	defer cancel()
+
+	log := a.Log.WithValues("loadtest", req.NamespacedName)
+
+	ctx, cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	// Fetch the LoadTest that triggers the event.
+	loadtest := new(grpcv1.LoadTest)
+	if err = a.Get(ctx, req.NamespacedName, loadtest); err != nil {
+		log.Error(err, "failed to get LoadTest")
+		// do not requeue, the load test may have been deleted
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If the triggering LoadTest is not yet terminated do nothing
+	if !loadtest.Status.State.IsTerminated() {
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch all the pods live on the cluster.
+	pods := new(corev1.PodList)
+	if err = a.List(ctx, pods, client.InNamespace(req.Namespace)); err != nil {
+		log.Error(err, "failed to list pods")
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Reuse existing logic to find the pods related to triggering LoadTest.
+	ownedPods := status.PodsForLoadTest(loadtest, pods.Items)
+
+	q := quitClient{}
+	quitWorkers(ctx, &q, ownedPods, log)
+
+	return ctrl.Result{}, nil
+}
+
+type quit interface {
+	callQuit(context.Context, *corev1.Pod, logr.Logger)
+}
+
+type quitClient struct {
+}
+
+// quitWorkers takes a list of pods and a log, check on each pod and send quit
+// RPC if the pod is a worker with status of running, pending and unknown.
+func quitWorkers(ctx context.Context, q quit, ownedPods []*corev1.Pod, log logr.Logger) {
+	for i := range ownedPods {
+		if ownedPods[i].Labels[config.RoleLabel] == config.DriverRole {
+			continue
+		}
+
+		if ownedPods[i].Status.Phase != corev1.PodFailed && ownedPods[i].Status.Phase != corev1.PodSucceeded {
```

Comment:
I think we should not rely on the PodPhase here b/c it is not always true. For example, many crashes are not marked as PodFailed. They may continue as pending with a reason string of "CrashLoopBackoff". I would rely on the status package's function to get the State for the Pod and use that. Then, we can update tests to mimic the conditions it looks for.

Reply:
I have updated to use `StateForPodStatus` method to check on status of the pods. Thank you for letting me know. :)

Reply:
Also update the test for this change.